### PR TITLE
OpenMRS Importer data types

### PR DESCRIPTION
**Pull request by kaapstorm**
Context: [IS-3715](https://infomovel.atlassian.net/browse/IS-3715)

##### SUMMARY
Currently importing patients from OpenMRS using `OpenmrsImporter` / via the OpenMRS Reporting API allows timestamps (represented as milliseconds since 1970-01-01) to be converted to ISO-formatted datetime strings.

This change brings data type conversion in line with the way the rest of MOTECH manages casting, while still supporting timezones. (Most importantly, this allows those timestamps to be converted to dates as well as datetimes.)

Easier to review :fish: :tropical_fish: :blowfish: 

##### FEATURE FLAG
"OpenMRS integration"

##### PRODUCT DESCRIPTION
When exporting data to OpenMRS, or when importing via the OpenMRS atom feed, MOTECH allows the data types of imported and exported values to be converted. This change extends that functionality to importing using OpenMRS Importers / via the OpenMRS Reporting API.

##### USAGE
To convert a value given in a report in milliseconds to a date, specify `"data_type": "posix_milliseconds"` and `"commcare_data_type": "cc_date"` when configuring the OpenMRS Importer. e.g.
```javascript
            {
                "column": "data_proxima_consulta",
                "data_type": "posix_milliseconds",
                "commcare_data_type": "cc_date",
                "property": "data_proxima_consulta"
            },  
```
To convert a decimal value to an integer, you only need to specify `"commcare_data_type": "cc_integer"`. e.g.
```javascript
             {
                "column": "filhos",
                "commcare_data_type": "cc_integer",
                "property": "numero_de_filhos"
            },
```

### CODE REVIEW

File: corehq/motech/openmrs/models.py [hunk 1]
```diff
@@ -9,33 +10,70 @@
...
+        local_serializers = serializers.copy()
+        local_serializers.update({
+            (OPENMRS_DATA_TYPE_MILLISECONDS, None): to_datetime_tz,
+            (OPENMRS_DATA_TYPE_MILLISECONDS, COMMCARE_DATA_TYPE_DATETIME): to_datetime_tz,
+            (OPENMRS_DATA_TYPE_MILLISECONDS, COMMCARE_DATA_TYPE_DATE): to_date_tz,
+        })
+
+        serializer = (local_serializers.get((self.data_type, self.commcare_data_type))
+                      or local_serializers.get((None, self.commcare_data_type)))
+        return serializer(external_value) if serializer else external_value
```
Comment (sravfeyn) [inline corehq/motech/openmrs/models.py]: Just curious, the title of the method is `deserialize` but the return values seem to be just serializers, how come?

File: corehq/motech/openmrs/models.py [hunk 2]
```diff
@@ -9,33 +10,70 @@
...
+
+    def _get_timezone(self):
+        return self._importer.get_timezone() if self._importer else None
+
+    def deserialize(self, external_value):
+        """
+        Returns ``external_value`` as its CommCare data type.
+        """
+        # Update serializers with timezone
+        to_datetime_tz = partial(omrs_timestamp_to_datetime, tz=self._get_timezone())
```
Comment (sravfeyn) [inline corehq/motech/openmrs/models.py]: If it's possible to pass timezone directly to this `deserialize` method, I think we can get rid of  `set_importer` and `_get_timezone` methods, unless there is a restriction on the method signature?

**Comment by kaapstorm** [comment]
@sravfeyn @millerdev do you reckon this is good to go yet?

**Review by sravfeyn — APPROVED** [review]

File: corehq/motech/openmrs/models.py [hunk 3]
```diff
@@ -37,6 +39,26 @@ class ColumnMapping(DocumentSchema):
     property = StringProperty()
     data_type = StringProperty(choices=DATA_TYPES, required=False)
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._importer = None
+
+    def set_importer(self, value):
+        self._importer = value
```
Comment (millerdev) [inline corehq/motech/openmrs/models.py]: Interesting use of a setter method. Why not use a plain attribute and assignment?

```py
def __init__(...)
    ...
    self.importer = None
```
and later
```py
mapping.importer = importer
```

Comment (kaapstorm) [reply corehq/motech/openmrs/models.py]: It's a jsonobject, so even though `importer` is an instance attribute, it'll still get returned by `to_json()`:
```python
In [12]: class Foo(DocumentSchema):
    ...:     foo = StringProperty()
    ...:     def __init__(self, *args, **kwargs):
    ...:         super().__init__(*args, **kwargs)
    ...:         self.bar = None
    ...:         

In [13]: foo = Foo()

In [14]: foo.to_json()
Out[14]: {'bar': None, 'doc_type': 'Foo', 'foo': None}
```

`ColumnMapping` instances are rendered in a JSON UI, so we would have to filter out the attributes we don't want users to see (or only select the ones we want them to see) in the view. It's messy. I figured a setter, although clumsy, would be less clumsy than the alternative.

**Pull request merged**